File: prover/datatypes.py
from __future__ import annotations
from typing import List, Dict, Tuple
from dataclasses import dataclass, field
from sexpdata import Symbol

__all__ = [
    "Name", "Term",
    "Lit", "Var", "Symtree", "Hole",
    "Derivation", "Sorry", "Proof",
    "InferenceRule", "State"
]

Name = str

class Term: pass

@dataclass
class Lit(Term):
    name : Name

@dataclass
class Var(Term):
    name : Name

@dataclass
class Symtree(Term):
    children : List[Term]

@dataclass
class Hole(Term):
    pass

def sexpr(τ):
    if isinstance(τ, Lit):
        return τ.name
    elif isinstance(τ, Var):
        return τ.name
    elif isinstance(τ, Symtree):
        return "(%s)" % " ".join(map(sexpr, τ.children))
    elif isinstance(τ, Hole):
        return "_"
    else:
        raise TypeError("this should never happens (naked Term)")
Term.__str__ = sexpr

class Derivation: pass

@dataclass
class Proof(Derivation):
    edge          : Name
    children      : List[Derivation]
    substitutions : Dict[Name, Term]

@dataclass
class Sorry(Derivation):
    name : str

@dataclass
class InferenceRule:
    premises   : List[Term]
    conclusion : Term

@dataclass
class State:
    variables : List[Name]               = field(default_factory=list)
    infix     : Dict[Symbol, int]        = field(default_factory=dict)
    context   : Dict[Name, State]        = field(default_factory=dict)
    bound     : List[Term]               = field(default_factory=list)
    defs      : List[Tuple[Term, Term]]  = field(default_factory=list)

File: ocaml/datatypes.ml
type sexp =
| Atom of string
| Int of int
| List of sexp list
| Supp of sexp list

let rec showSExp : sexp -> string = function
  | Atom s  -> s
  | Int x   -> string_of_int x
  | List xs -> "(" ^ String.concat " " (List.map showSExp xs) ^ ")"
  | Supp xs -> "[" ^ String.concat " " (List.map showSExp xs) ^ "]"

type name = string

type term =
| Lit of name
| Var of name
| Symtree of term list
| Hole

let rec showTerm : term -> string = function
  | Lit s -> s
  | Var s -> s
  | Symtree xs -> "(" ^ String.concat " " (List.map showTerm xs) ^ ")"
  | Hole -> "_"

module Env = Map.Make(String)

type env = term Env.t

type argument =
  | Sorry of name
  | Lemma of name

type proof =
{ edge          : name;
  arguments     : argument list;
  substitutions : env }

type inferenceRule =
{ premises   : term list;
  conclusion : term }

type state =
{ variables : name list;
  infix     : int Env.t;
  context   : inferenceRule Env.t;
  bound     : term list;
  defs      : (term * term) list }